### PR TITLE
Fix chef provisioner validateFn

### PR DESCRIPTION
**Pull request by jeremiahsnapp**
Correctly validate Chef provisioner's `use_policyfile` field even if its value is a string type.

This fixes https://github.com/hashicorp/terraform/issues/16655

### CODE REVIEW

**Comment by svanharmelen** [comment]
Thanks for your PR @jeremiahsnapp! I see what you are trying to do, but I wonder if this isn't actually a bug in Terraform core instead.

@apparentlymart the [schema](https://github.com/jeremiahsnapp/terraform/blob/b363193261f9d1fe049167f1e4fe3e81eeadf894/builtin/provisioners/chef/resource_provisioner.go#L175-L178) clearly specifies the `use_policyfile` field as being a `schema.TypeBool`, so shouldn't TF core return an error on this instead?

**Comment by jeremiahsnapp** [comment]
Thanks for looking at this @svanharmelen and @apparentlymart. I'm new to the terraform code so I'm not certain if their is a bug in the core code. I will say that the [Booleans docs](https://www.terraform.io/docs/configuration/variables.html#booleans) gives me the impression that using string values for a boolean variable is preferred and that TF should automatically convert the value internally to a boolean as needed.

The [chef provisioner's validateFn function](https://github.com/jeremiahsnapp/terraform/blob/v0.11.2/builtin/provisioners/chef/resource_provisioner.go#L372) is being passed a `terraform.ResourceConfig`. I believe TF's core [ResourceConfig.Get](https://github.com/hashicorp/terraform/blob/v0.11.2/terraform/resource.go#L232) _does not_ do the automatic conversion from string to boolean because it _does not_ take into account the data's schema.

Contrast this with how the [chef provisioner's decodeConfig function](https://github.com/jeremiahsnapp/terraform/blob/v0.11.2/builtin/provisioners/chef/resource_provisioner.go#L726) is passed a `schema.ResourceData`. I believe the [ResourceData.Get](https://github.com/hashicorp/terraform/blob/v0.11.2/helper/schema/resource_data.go#L70) _does_ do the automatic conversion from string to boolean because it _does_ take into account the data's schema.

So while the `schema.ResourceData` part of the code is auto converting the variable properly the `terraform.ResourceConfig` part is not. If that is intentional by design then I think my PR is required to fix the chef provisioner. If that is not the intended design then I agree that the bug is somewhere in TF core. Hopefully one of you can help isolate this issue better.

Thanks!

File: builtin/provisioners/chef/resource_provisioner.go [hunk 1]
```diff
@@ -370,18 +371,27 @@ func applyFn(ctx context.Context) error {
...
+		switch usePolicyFileRaw := usePolicyFileRaw.(type) {
+		case string:
+			if usePolicyFileBool, err := strconv.ParseBool(usePolicyFileRaw); err == nil {
+				usePolicyFile = usePolicyFileBool
+			} else {
+				es = append(es, errors.New("\"use_policyfile\": field is not a boolean value"))
+			}
+		case bool:
+			usePolicyFile = usePolicyFileRaw
+		}
```
Comment (svanharmelen) [inline builtin/provisioners/chef/resource_provisioner.go]: May I suggest a small rewrite of your `switch` statement to make it just a bit more idiomatic Go?

```go
switch usePolicyFileRaw := usePolicyFileRaw.(type) {
  case bool:
    usePolicyFile = usePolicyFileRaw
  case string:
    usePolicyFileBool, err := strconv.ParseBool(usePolicyFileRaw)
    if err != nil {
      return append(es, errors.New("\"use_policyfile\" must be a boolean"))
    }
    usePolicyFile = usePolicyFileBool
  default:
    return append(es, errors.New("\"use_policyfile\" must be a boolean")) 
}
```

**Comment by jeremiahsnapp** [comment]
Thanks a lot for the idiomatic Go advice @svanharmelen. I like it a lot. I only made a minor change to your return statements so they return the correct number of arguments. Thanks!

**Comment by svanharmelen** [comment]
Again thanks for the PR and your inputs! LGTM! 🎉

**Pull request merged**